olive/passes/onnx/peephole_optimizer.py

```diff
@@ -232,13 +235,142 @@ def fuse_reshape_operations(self):
+
+    def _find_value_info_proto_by_name(self, model, name):
+        """Find the ValueInfoProto with the name name in the model's value_info."""
+        for vi in model.graph.value_info:
+            if vi.name == name:
+                return vi
+
+        for initializer in model.graph.initializer:
+            if initializer.name == name:
+                return helper.make_tensor_value_info(name, initializer.data_type, initializer.dims)
+
+        for graph_input in model.graph.input:
+            if graph_input.name == name:
+                return graph_input
+
+        raise ValueError(f"No value info proto named {name}")
+
+    def _run_op(self, model, op):
+        input_names = set()
+
+        op_model = onnx.ModelProto()
+        op_model.ir_version = model.ir_version
+        op_model.producer_name = "constant_folding"
+        op_model.opset_import.extend(model.opset_import)
+        op_model.graph.name = "ConstantFoldingGraph"
+        op_model.graph.node.extend([copy.deepcopy(op)])
+
+        for input_name in op.input:
+            if input_name and input_name not in input_names:
+                try:
+                    initializer = self._find_initializer_by_name(model, input_name)
+                    op_model.graph.initializer.append(copy.deepcopy(initializer))
+                    vi = helper.make_tensor_value_info(initializer.name, initializer.data_type, initializer.dims)
+                    op_model.graph.input.append(vi)
+                except ValueError:
+                    vi = self._find_value_info_proto_by_name(model, input_name)
+                    op_model.graph.input.append(copy.deepcopy(vi))
+                input_names.add(input_name)
+
+        for output_name in op.output:
+            vi = helper.make_tensor_value_info(output_name, TensorProto.UNDEFINED, [])
+            op_model.graph.output.append(vi)
+
+        return op_model
+
+    def _run_onnx_model(self, model):
+        session = ort.InferenceSession(model.SerializeToString(), providers=["CPUExecutionProvider"])
+        input_dict = {}
+        for model_input in session.get_inputs():
+            name = model_input.name
+            tensor = self._find_initializer_by_name(model, name)
+            input_dict[name] = numpy_helper.to_array(tensor)
+        return session.run(None, input_dict)
+
+    def _get_constant_nodes(self, model):
+        dynamic_inputs = {graph_input.name for graph_input in model.graph.input}
+        const_inputs = {
+            initializer.name for initializer in model.graph.initializer if initializer.name not in dynamic_inputs
+        }
+        const_nodes = []
+        for node in model.graph.node:
+            if all(node_input == "" or node_input in const_inputs for node_input in node.input):
+                const_nodes.append(node)
+                const_inputs.update(node.output)
+        return const_nodes
+
+    def fold_constant(self):
+        model_copy = copy.deepcopy(self.model)
```

Comment:
is there a reason to copy the model? what if the changes are done directly on the original model? self.model is not touched again until assignment with model_copy.

Reply:
This logic modifies the model by adding new initializers and removing nodes. Using a deep copy ensures that if any error occurs during this process, the original model remains unchanged.

Reply:
Are you worried about the input model being affected by this or something? There is no caching of the model proto like we do for loaded pytorch model so that's not an issue. 

If there is an error here, then the pass will fail. this behavior does not change whether we change the original model proto or a copy of it.

Reply:
That's fair, i originally thought it has try-catch block so the logic may fail, but the behavior should be same. I'll remove this copy part thanks!